tests/e2e/http_health_check_test.go

```diff
@@ -0,0 +1,211 @@
+					url:                "/health?serializable=false",
+					expectTimeoutError: true,
+				},
+			},
+		},
+		{
+			name:           "raft loop deadlock auth is enabled",
+			injectFailure:  triggerRaftLoopDeadLock,
+			clusterOptions: []e2e.EPClusterOption{e2e.WithClusterSize(3), e2e.WithGoFailEnabled(true)},
+			authEnabled:    true,
+			healthChecks: []healthCheckConfig{
+				{
+					// current kubeadm etcd liveness check failed to detect raft loop deadlock in steady state
+					url:              "/health?serializable=true",
+					expectStatusCode: http.StatusOK,
+				},
+				{
+					url:                "/health?serializable=false",
+					expectTimeoutError: true,
+				},
+			},
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			clus, err := e2e.NewEtcdProcessCluster(ctx, t, tc.clusterOptions...)
+			require.NoError(t, err)
+			defer clus.Close()
+			testutils.ExecuteUntil(ctx, t, func() {
+				if tc.authEnabled {
+					setupAuth(ctx, t, clus.Etcdctl())
+				}
+				tc.injectFailure(ctx, t, clus)
+
+				for _, hc := range tc.healthChecks {
+					requestURL := clus.Procs[0].EndpointsHTTP()[0] + hc.url
+					t.Logf("health check URL is %s", requestURL)
+					doHealthCheckAndVerify(t, client, requestURL, hc.expectStatusCode, hc.expectTimeoutError)
+				}
+			})
+		})
+	}
+}
+
+func doHealthCheckAndVerify(t *testing.T, client *http.Client, url string, expectStatusCode int, expectTimeoutError bool) {
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	require.NoErrorf(t, err, "failed to creat request %+v", err)
+	resp, herr := client.Do(req)
+	cancel()
+	if expectTimeoutError {
+		require.Contains(t, herr.Error(), context.DeadlineExceeded.Error())
+		return
+	}
+	require.NoErrorf(t, herr, "failed to get response %+v", err)
+
+	body, err := io.ReadAll(resp.Body)
+	resp.Body.Close()
+	require.NoErrorf(t, err, "failed to read response %+v", err)
+
+	t.Logf("health check response body is: %s", body)
+	require.Equal(t, expectStatusCode, resp.StatusCode)
+}
+
+func noop(_ context.Context, _ *testing.T, _ *e2e.EtcdProcessCluster) {}
+
+func triggerNoSpaceAlarm(ctx context.Context, t *testing.T, clus *e2e.EtcdProcessCluster) {
+	buf := strings.Repeat("b", os.Getpagesize())
+	etcdctl := clus.Etcdctl()
+	for {
+		if err := etcdctl.Put(ctx, "foo", buf, config.PutOptions{}); err != nil {
+			if !strings.Contains(err.Error(), "etcdserver: mvcc: database space exceeded") {
+				t.Fatal(err)
+			}
+			break
+		}
+	}
+}
+
+func triggerSlowApply(ctx context.Context, t *testing.T, clus *e2e.EtcdProcessCluster) {
+	// the following proposal will be blocked at applying stage
+	// because when apply index < committed index, linearizable read would time out.
+	require.NoError(t, clus.Procs[0].Failpoints().Setup(ctx, "beforeApplyEntryNormal", `sleep(3000)`))
```

Comment:
Did you mean 3 seconds?
```suggestion
	require.NoError(t, clus.Procs[0].Failpoints().Setup(ctx, "beforeApplyEntryNormal", `sleep(3s)`))
```